extensions/sockets/src/main/java/org/mule/module/socket/internal/worker/UdpWorker.java

```diff
@@ -0,0 +1,108 @@
+import static org.mule.module.socket.internal.SocketUtils.createMuleMessage;
+import static org.mule.module.socket.internal.SocketUtils.createPacket;
+import static org.mule.module.socket.internal.SocketUtils.getUdpAllowedByteArray;
+import org.mule.module.socket.api.source.ImmutableSocketAttributes;
+import org.mule.module.socket.api.source.SocketAttributes;
+import org.mule.runtime.api.execution.CompletionHandler;
+import org.mule.runtime.api.execution.ExceptionCallback;
+import org.mule.runtime.api.message.MuleEvent;
+import org.mule.runtime.core.api.MuleContext;
+import org.mule.runtime.core.api.serialization.ObjectSerializer;
+import org.mule.runtime.extension.api.runtime.MessageHandler;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * One worker is created per received package. If the other end of the connection is awaiting
+ * for a response, one will be sent but not from the same listener socket the source has. The response
+ * will be sent from a new different {@link DatagramSocket} bound to a port choose by the system.
+ */
+public final class UdpWorker extends SocketWorker
+{
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UdpWorker.class);
+    private final DatagramSocket socket;
+    private final DatagramPacket packet;
+    private final ObjectSerializer objectSerializer;
+
+    public UdpWorker(DatagramSocket socket, DatagramPacket packet, ObjectSerializer objectSerializer, MuleContext muleContext, MessageHandler<InputStream, SocketAttributes> messageHandler)
+    {
+        super(muleContext, messageHandler);
+        this.socket = socket;
+        this.packet = packet;
+        this.objectSerializer = objectSerializer;
+    }
+
+    @Override
+    public void run()
+    {
+        SocketAttributes attributes = new ImmutableSocketAttributes(packet);
+        InputStream content = new ByteArrayInputStream(copyOf(packet.getData(), packet.getLength()));
+        messageHandler.handle(createMuleMessage(content, attributes, muleContext), new CompletionHandler<MuleEvent, Exception, MuleEvent>()
+        {
+            @Override
+            public void onCompletion(MuleEvent muleEvent, ExceptionCallback<MuleEvent, Exception> exceptionCallback)
+            {
+                try
+                {
+                    byte[] byteArray = getUdpAllowedByteArray(muleEvent.getMessage().getPayload(), objectSerializer);
+                    DatagramPacket sendPacket = createPacket(byteArray);
+                    sendPacket.setSocketAddress(packet.getSocketAddress());
+                    socket.send(sendPacket);
+                }
+                catch (IOException e)
+                {
+                    LOGGER.error(format("An error occurred while sending UDP packet to address '%s'", packet.getSocketAddress().toString()));
```

Comment:
there's no point in logging all of this... just create an exception which has the cause and this message and pass it to the exceptionCallback